### PR TITLE
Add support for timestamp type to Pinot connector

### PR DESCRIPTION
**Pull request by ddcprg**
## Description
Add support for TIMESTAMP data type in Pinot connector. Help fixing https://github.com/trinodb/trino/issues/10199

> Is this change a fix, improvement, new feature, refactoring, or other?
Improvement

> Is this a change to the core query engine, a connector, client library, or the SPI interfaces? (be specific)
Pinot connector

> How would you describe this change to a non-technical end user or system administrator?
Include additional Pinot data types in Pinot connector

## Documentation

( ) No documentation is needed.
(X) Sufficient documentation is included in this PR.
( ) Documentation PR is available with #prnumber.
( ) Documentation issue #issuenumber is filed, and can be handled later.

## Release notes

( ) No release notes entries required.
(X) Release notes entries required with the following suggested text:

```markdown
# Section
* Add support for TIMESTAMP data type in Pinot connector. ({issue}`10199`)
```

### CODE REVIEW

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java
Comment (ddcprg) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java:35]: See https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java#L468

File: plugin/trino-pinot/src/test/resources/alltypes_schema.json
Comment (ddcprg) [inline plugin/trino-pinot/src/test/resources/alltypes_schema.json:47]: I'll add one more test column for multi-value timestamps

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (ddcprg) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:137]: not sure if this is correct. `TimestampType.TIMESTAMP` is deprecated and set to `TimestampType.TIMESTAMP_MILLIS`

**Comment by hashhar** [comment]
@elonazoulay can you please help with an initial review? I won't be able to review this thoroughly until ~9th May so expect some delays.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (xiangfu0) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: Should this be JsonType or VarcharType? How JSON functions will be handled here e.g. json_extract ?

Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: I was going to leave functions out of the scope of this PR but I'll check how easy it would be to include them otherwise I'll raise a separate issue for this. Probably the only 2 function that could be easily mapped are `json_extract_scalar` and `json_extract_scalar` as each has an equivalent in Pinot. I'll get back to you on this

**Comment by nizarhejazi** [comment]
@elonazoulay can you please help with PR review?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: I see, can you test it locally with [PinotJsonQuickStart](https://github.com/apache/pinot/blob/master/pinot-tools/src/main/java/org/apache/pinot/tools/JsonIndexQuickStart.java)?
Here is a reference PR as well.
https://github.com/prestodb/presto/pull/17015/files#diff-379277f1690c545f2f182815aa2b984399e26318df7e2402be7149a2c5e035b4R135

https://github.com/prestodb/presto/pull/17323

Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: Thanks @xiangfu0 I'll take a look at those. In the meantime it seems I've found an issue with multi-value timestamp columns, I've raised an issue in the pinot project https://github.com/apache/pinot/issues/8624

**Comment by nizarhejazi** [comment]
@ddcprg thanks for putting this together. @hashhar @elonazoulay can we do another round of reviews on Monday? Thanks.

**Comment by nizarhejazi** [comment]
@hashhar @elonazoulay @xiangfu0 Can you review PR and sign off if no blocker?

**Comment by hashhar** [comment]
I'll take a look during the end of this week.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java:55]: Shouldn't we throw if we still receive `null` here?

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:52]: Avoid casts from `Number`. `Byte`, `BigDecimal` etc. are all subclasses of `Number`.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (hashhar) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:146]: inline since they are used once and I make the test data appear closer with all the other test data.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (hashhar) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1564]: Does this return correct results if both a `null` and `1970-01-01T00:00:00` exist?

**Comment by nizarhejazi** [comment]
@ddcprg Can you address @hashhar feedback? thanks.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (ddcprg) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1564]: the default null value for timestamps is epoch so this should return `1970-01-01T00:00:00`

**Comment by nizarhejazi** [comment]
@hashhar @xiangfu0 can you take another look? thanks.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (xiangfu0) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:61]: Why write microseconds to this TIMESTAMP_MILLIS?

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (hashhar) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1564]: Then I would argue it's unsafe to push-down aggregates in presence of `null`s. Maybe we can instead use query-passthrough (by implementing a polymorphic table function as in https://github.com/trinodb/trino/pull/12325) for such cases to get Pinot semantics.

cc: @findepi Any opinions? i.e. DISTINCT on a column with `null` and `epoch` would return `epoch` since the `null` becomes `epoch`.

**Comment by nizarhejazi** [comment]
@ddcprg any update on this?

**Comment by nizarhejazi** [comment]
@ddcprg @xiangfu0 @hashhar @findepi any update on this PR? It has been in review for more than a month and half!

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:61]: @nizarhejazi @ddcprg can you check this?
 Also make the CI passing?

Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:61]: @xiangfu0 I'm absent at the moment. I'll retake this in 2 weeks time

**Comment by ddcprg** [comment]
@nizarhejazi I'm absent at the moment. I'll retake this in 2 weeks time

**Comment by xiangfu0** [comment]
@ddcprg, are you back on this ? Can you try to rebase and fix the test? Thanks!

**Review by xiangfu0 — APPROVED** [review]

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/conversion/PinotTimestamps.java:29]: This should be: 
DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss[.SSS][.SS][.S]")

**Comment by ddcprg** [comment]
@xiangfu0 I'll take a look at the comments today, thanks for your feedback

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java]: @hashhar @xiangfu0 Pinot's MAX function returns double https://docs.pinot.apache.org/users/user-guide-query/supported-aggregations I think the options are either keeping `Number` or having 2 conditionals: one for `Double` and one for `Long`. I rather keep checking for `Number` type, let me know your thoughts

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (ddcprg) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1574]: I'm going to double check and get back to you. I think this is the same behaviour of other types like long and double

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (xiangfu0) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:53]: I think number is ok.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (xiangfu0) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1574]: yes, Pinot schema will assign the default null value. So this behavior is expected.

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (xiangfu0) [inline plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1174]: Can you also add a test for timestamp predicate and pushdown?
E.g. `where timestamp_col > timestamp '2022-01-01 00:00:00.000'`

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (ebyhr) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: You can get `JsonType` via `TypeManager`. `PostgreSqlClient` has the example.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (ebyhr) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:137]: `TimestampType.TIMESTAMP_MILLIS` is correct.

**Comment by ddcprg** [comment]
test

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (ddcprg) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1174]: @xiangfu0 thanks for suggesting this test! I found I had to add one more conversion

**Comment by ebyhr** [comment]
> is it fine if we keep it like this for now? 

I would defer to @hashhar (I haven't reviewed this PR at all)

**Comment by hashhar** [comment]
Sorry for the delay here. I'll take a look over the weekend.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java:170]: only for ShortTimestampType i.e. precision <= 6.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java
Comment (hashhar) [inline plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java:260]: Does Pinot only support timestamps with precision <= 6? If not we should probably add a `verify` here to make sure we don't silently do incorrect things with higher precision values.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:52]: having `else if (value instanceof Double || value instanceof Long)` is still useful in my opinion.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (hashhar) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:61]: reminder

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (hashhar) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1581]: We shouldn't push down such queries then probably. While the results match semantics of Pinot they don't match Trino so anyone performing a JOIN or SELECT across two catalogs will see inconsistent results.

For preserving Pinot semantics we should implement something like https://trino.io/docs/current/connector/postgresql.html#query-varchar-table instead.

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java:260]: pinot timestamps are milliseconds, see https://github.com/apache/pinot/blob/master/pinot-spi/src/main/java/org/apache/pinot/spi/data/FieldSpec.java#L468

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotSegmentPageSource.java:170]: I'll make an additional note

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:52]: this is a temporary conversion to micros to be able invoke `TIMESTAMP_MILLIS.writeLong(output, epochMicros);` if there is a better way to write milliseconds to the output block please let me know. Please read the javadocs in `ShortTimestampType`

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:61]: this is a temporary conversion to micros to be able invoke `TIMESTAMP_MILLIS.writeLong(output, epochMicros);` if there is a better way to write milliseconds to the output block please let me know. Please read the javadocs in `ShortTimestampType`

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (ddcprg) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1581]: in pinot the only way to make distinction between null and non-null values is by using `IS NULL` and `IS NOT NULL` other than this pinot will assign the default value if null. Push-down already happens for other types, should this be addressed in a separate ticket? I could take a look at this once we merge this PR. I'll summon @elonazoulay on this one

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java
Comment (ddcprg) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/decoders/TimestampDecoder.java:52]: I've now made this change

File: plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java
Comment (ebyhr) [reply plugin/trino-pinot/src/main/java/io/trino/plugin/pinot/PinotColumnHandle.java:132]: I think it would be better to revert this change and implement in #13428

**Comment by elonazoulay** [comment]
Hi @ddcprg ! I have a pr, we use json support internally here for some time but I just pushed this, didn't realize you are doing something similar: https://github.com/trinodb/trino/pull/13428 - would it make sense to just implement the timestamp support in this pr? We have a really good use case for it but did not implement it. lmk if you want to chat offline, I can help test this as well. Also if you could take a look at #13428 whenever you have some time. lmk what works for you.

**Comment by ddcprg** [comment]
@elonazoulay @ebyhr I'm happy to remove the JSON type from this PR, you done a lot of refactoring in the other PR which what was initially avoiding with this one. This PR has been open for long time though so wonder if we just drop it, unfortunately I don't have plenty of time to dedicate to open source contributions

**Comment by ddcprg** [comment]
@elonazoulay is a bit late for me right now, I'll ping you over the chat tomorrow

**Comment by xiangfu0** [comment]
+1 on separated the JSON and TIMESTAMP type support

**Comment by ddcprg** [comment]
@xiangfu0  @elonazoulay  @ebyhr @hashhar the scope of this PR is now reduced to simple timestamp types, please take a look when you have a min

**Review by xiangfu0 — APPROVED** [review]
lgtm

File: plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java
Comment (elonazoulay) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1549]: I tried this for filter pushdown - the behavior is not consistent with trino sql: you can select ... where is null and the result will be non null (i.e. the default value). Will take a look and see what can be done.

Comment (ddcprg) [reply plugin/trino-pinot/src/test/java/io/trino/plugin/pinot/AbstractPinotIntegrationSmokeTest.java:1549]: yes, the projection in Pinot will always return the default value if column is null

**Comment by etadelta222** [comment]
Just wanted to follow-up on this 👀

**Comment by jatin5251** [comment]
@ddcprg can you please resolve conflicts if possible then we can request for merging thanks

**Comment by xiangfu0** [comment]
> @ddcprg can you please resolve conflicts if possible then we can request for merging thanks

+1

**Comment by xiangfu0** [comment]
hi @ddcprg, could you please resolve the conflict?

**Comment by xiangfu0** [comment]
Since I cannot edit @ddcprg 's branch to amend this PR. I took his branch and rebased the code to latest master and create a new PR: https://github.com/trinodb/trino/pull/14244

**Pull request closed without merging**
